[test] More deserialization recovery vtable layout tests

jrose-apple opened the pull request:
rdar://problem/35508966

jrose-apple committed:
[test] Add deserialization recovery test for 'override required init'

This could have had different behavior from normal 'required', but it
doesn't. Everything seems to work.

jrose-apple commented:
@slavapestov Anything else you can think of?

@swift-ci Please test

slavapestov commented on test/Serialization/Recovery/typedefs.swift:
```diff
@@ -144,7 +144,7 @@ open class User {
   public subscript(_: WrappedInt) -> () { return () }
 
-  // CHECK: subscript<T>(_: T) -> () where T : HasAssoc, T.Assoc == WrappedInt { get }
+  // CHECKx: subscript<T>(_: T) -> () where T : HasAssoc, T.Assoc == WrappedInt { get }
```
Is CHECKx intentional?

jrose-apple replied:
Bah, I hadn't rebuilt everything to pick up Nathan's fix for this. Thanks for the catch.

jrose-apple committed:
[test] Add deserialization recovery test for 'dynamic required'

We already a section on this but it didn't check vtable offsets.
Better to have this explicitly tested!

rdar://problem/35508966

jrose-apple commented:
@swift-ci Please smoke test and merge

jrose-apple commented:
@swift-ci Please smoke test and merge
